src/admin/class-cmb2.php

```diff
@@ -107,25 +109,46 @@ public static function cmb2_render_markerpreview_field_type( $field, $escaped_va
 	 * @return string The field type HTML.
 	 */
 	public static function cmb2_render_openstreetmap_field_type( $field, $escaped_value, $object_id ) {
-		$args = wp_parse_args(
-			[
-				'post_type'   => 'owc_ok_location',
-				'numberposts' => -1,
-				'orderby'     => 'title',
-				'order'       => 'ASC',
-				// phpcs:ignore WordPress.DB.SlowDBQuery -- This query is needed to get the locations.
-				'meta_query'  => [
-					'relation' => 'OR',
+
+		$datalayer_type = get_post_meta( $object_id, 'datalayer_type', true );
+
+		switch ( $datalayer_type ) {
+			case 'fileinput':
+			case 'url':
+			default:
+				$meta_query = [
+					'relation' => 'AND',
 					[
-						'key'     => 'location_datalayer_id',
-						'value'   => $object_id,
-						'compare' => '=',
+						'key'     => 'geometry',
+						'compare' => 'EXISTS',
 					],
-				],
-			]
-		);
+				];
 
-		$datalayer_locations = get_posts( $args );
+				$meta_query = array_merge(
+					$meta_query,
+					[
+						[
+							'key'     => 'location_datalayer_id',
+							'value'   => $object_id,
+							'compare' => '=',
+						],
+					]
+				);
+
+				$args = wp_parse_args(
+					[
+						'post_type'   => 'owc_ok_location',
+						'numberposts' => -1,
+						'orderby'     => 'title',
+						'order'       => 'ASC',
+						// phpcs:ignore WordPress.DB.SlowDBQuery -- This query is needed to get the locations.
```

Comment:
Volgens mij gebruik je verderop in de code alleen de id's van de posts, dus dan zou je hier ook alleen de id's kunnen ophalen: `'fields' => 'ids'`, maar wellicht is het dan zelfs beter om een custom query te gebruiken, dat is waar de SlowDBQuery sniff immers over gaat.

Reply:
Ik heb er een custom query van gemaakt, waar ik alleen het ID ophaal